### PR TITLE
Add normalization and verification of axis for AMDQuarkBFPQuantizeDequantizeOp

### PR DESCRIPTION
**Pull request by jorickert**
The axis in the model can be negative

### CODE REVIEW

File: src/Dialect/ONNX/ONNXOps/Additional/AMDQuark.cpp [hunk 1]
```diff
@@ -30,6 +30,14 @@ LogicalResult AMDQuarkBFPQuantizeDequantizeOp::verify() {
...
   }
 
+  if (auto rankedType = dyn_cast<RankedTensorType>(getX().getType())) {
+    const int64_t rank = rankedType.getRank();
+    const int64_t axis = getAxis();
+    if (axis < -rank || axis >= rank)
+      return emitOpError("axis attribute value ")
+             << axis << " is out of range [-" << rank << ", " << rank << ")";
+  }
+
```
Comment (gerion-amd) [inline src/Dialect/ONNX/ONNXOps/Additional/AMDQuark.cpp:40]: Would it make sense to put this code in `getNormalizedAxis()` and maybe just call this function here (I'm currently not familiar with when `verify()` is called but I would expect from `getNormalizedAxis` to always return valid axes.)

File: src/Dialect/ONNX/ONNXOps/Additional/AMDQuark.cpp [hunk 2]
```diff
@@ -48,8 +56,13 @@ struct KnownConfig {
...
   if (op->getBfpMethod() != config.method)
     return false;
-  if (!ignoreAxis && op->getAxis() != config.axis)
-    return false;
+  if (!ignoreAxis) {
+    const auto normalizedAxis = op->getNormalizedAxis();
+    if (!normalizedAxis)
+      return false;
+    if (int64_t(*normalizedAxis) != config.axis)
+      return false;
```
Comment (gerion-amd) [inline src/Dialect/ONNX/ONNXOps/Additional/AMDQuark.cpp:62]: I think, this works:
```
if (!normalizedAxis || int64_t(*normalizedAxis) != config.axis)
  return false;
```

File: test/mlir/onnx/invalid.mlir [hunk 3]
```diff
@@ -1024,3 +1024,19 @@ func.func @test_bfp_quant_dequant_wrong_rounding_mode(%arg0: tensor<16x32xf32>)
...
+  return %0 : tensor<16x32xf32>
+}
+
+// -----
+
+func.func @test_bfp_quant_dequant_axis_negative_out_of_range(%arg0: tensor<16x32xf32>) -> tensor<16x32xf32> {
+  // expected-error @+1 {{'onnx.AMDQuarkBFPQuantizeDequantizeOp' op axis attribute value -3 is out of range [-2, 2)}}
+  %0 = "onnx.AMDQuarkBFPQuantizeDequantizeOp"(%arg0) { axis = -3: si64 }  : (tensor<16x32xf32>) -> tensor<16x32xf32>
+  return %0 : tensor<16x32xf32>
+}
```
Comment (gerion-amd) [inline test/mlir/onnx/invalid.mlir:1042]: Please also add a test for a valid negative axis.

**Review by gerion-amd — APPROVED** [review]

**Review by mathmer-amd — APPROVED** [review]
Can you add a test where the axis is negative but inside the range please?

File: src/Dialect/ONNX/ONNXOps/Additional/AMDQuark.cpp (hunk 1 above)
Comment (jorickert) [reply src/Dialect/ONNX/ONNXOps/Additional/AMDQuark.cpp:40]: verify() is called in multi places during the MLIR flow, depending on flags used, for example after op construction or after pattern/pass application.

To simlify the code a bit, i removed the assert in getNormalizedAxis. If the op is valid, getNormalizedAxis will also return a valid axis. If its not, the verify() should catch it before.
